src/lib/operators/join_hash.cpp

```diff
@@ -698,55 +722,91 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
 
       // we need to swap back the inputs, so that the order of the output columns is not harmed
       if (_inputs_swapped) {
-        write_output_chunks(output_chunk, _right_in_table, right, ref_col_right);
+        write_output_chunks(output_chunk, _right_in_table, right_pos_lists_by_column, right, ref_col_right);
 
         // Semi/Anti joins are always swapped but do not need the outer relation
-        if (_mode != JoinMode::Semi && _mode != JoinMode::Anti) {
-          write_output_chunks(output_chunk, _left_in_table, left, ref_col_left);
+        if (only_output_right_input) {
+          write_output_chunks(output_chunk, _left_in_table, left_pos_lists_by_column, left, ref_col_left);
         }
       } else {
-        write_output_chunks(output_chunk, _left_in_table, left, ref_col_left);
-        write_output_chunks(output_chunk, _right_in_table, right, ref_col_right);
+        write_output_chunks(output_chunk, _left_in_table, left_pos_lists_by_column, left, ref_col_left);
+        write_output_chunks(output_chunk, _right_in_table, right_pos_lists_by_column, right, ref_col_right);
       }
       _output_table->emplace_chunk(std::move(output_chunk));
     }
 
     return _output_table;
   }
 
+  // See usage in _on_execute() for doc.
+  static PosListsByColumn setup_pos_lists_by_column(const std::shared_ptr<const Table> input_table) {
+    std::map<PosLists, std::shared_ptr<PosLists>> shared_pos_lists_by_pos_lists;
+
+    PosListsByColumn pos_lists_by_column(input_table->column_count());
+    auto pos_lists_by_column_it = pos_lists_by_column.begin();
+
+    const auto& input_chunks = input_table->chunks();
+
+    for (ColumnID column_id{0}; column_id < input_table->column_count(); ++column_id) {
+      // Get all the input pos lists so that we only have to pointer cast the columns once
+      auto pos_list_ptrs = std::make_shared<PosLists>(input_table->chunk_count());
+      auto pos_lists_iter = pos_list_ptrs->begin();
+
+      for (ChunkID chunk_id{0}; chunk_id < input_table->chunk_count(); chunk_id++) {
+        // This works because we assume that the columns have to be either all ReferenceColumns or none.
```

Comment:
... and we have checked that the first one is a reference column before we called this method.

Reply:
Added assert "DebugAssert(input_table->get_type() == TableType::References, "Function only works for reference tables");" and removed the comment